### PR TITLE
Feature/zero allocation render loop 104794834

### PR DESCRIPTION
**Pull request by Kuranes**
Prevent any repeating matrix allocation each frame by keeping stack whenever it makes sense or just plain removes it when not needed

### CODE REVIEW

File: sources/osg/Matrix.js [hunk 1]
```diff
@@ -1223,6 +1223,34 @@ define( [
             };
         } )(),
 
+        scale: function ( x, y, z, m, res ) {
+            if ( res === undefined ) {
```
Comment (stephomi) [inline sources/osg/Matrix.js]: !res ?

File: sources/osg/Matrix.js [hunk 2]
```diff
@@ -1223,6 +1223,34 @@ define( [
...
+                Notify.warn( 'no matrix destination !' );
+                res = Matrix.create();
+            }
+            // fill CPU cache L1
+            Matrix.copy( m, res );
```
Comment (stephomi) [inline sources/osg/Matrix.js]: I think it's best if we stick with the glmatrix version...
Plus... we are writing twice in the res array (and we are reading twice the same values too, once in the copy call and once again with the ``res[ x ] *=`` ).

File: sources/osg/Node.js [hunk 3]
```diff
@@ -341,10 +365,41 @@ define( [
...
+
+            var m = Node.__reserveMatrixStack[ Node._reserveMatrixStack.current++ ];
+
+            if ( Node._reserveMatrixStack.current === Node._reserveMatrixStack.length ) {
+
+                Node._reserveMatrixStack.push( Matrix.create() );
+
+            } else {
+
+                Matrix.setIdentity( m );
```
Comment (stephomi) [inline sources/osg/Node.js]: makeIdentity ?...

**Comment by Kuranes** [comment]
Question left is MatrixClass for this PR or another one ?

**Comment by cedricpinson** [comment]
It would be a good candidate to do a MatrixPool now I think, if you want we can talk about it

**Comment by Kuranes** [comment]
ok.

**Comment by cedricpinson** [comment]
Maybe MatrixPool is a better name ?

**Comment by Kuranes** [comment]
MatrixMemoryPool then would be clearer as Pool might not be that clear ?
( indeed matrixstack can confuse with other matrix stack in osg )

**Comment by cedricpinson** [comment]
needs instructions to test and demonstrate the improvement

**Comment by Kuranes** [comment]
Yes, putting back need fixing as there ismultiple improvements  and each need scene test to be build by code, and some fairly complex to test.

For now, juste have the testing on frustum Example with ?count=25&starts=1 shows a Culltime from 45 to 30ms. which shows clearly a matrixstack usage improvement of computeWorldMatrix martrixmemorypool.

For the others
- Shadow Map improvements more complex, have to make a test scene with many nodes + shadow to see the improvements on computebounds. 
- StackedScale change need animations with many scale track to  test

**Comment by cedricpinson** [comment]
We dont need a full report on all example just some sample scene built with code that can be executed easily

**Comment by Kuranes** [comment]
```
Perf:                Before    After
Basic                254        140
Cull                 179        104
Shadow               425        408
```

**Comment by cedricpinson** [comment]
What's the number means ? how do you get them ?

File: tests/osg/Matrix.js [hunk 4]
```diff
@@ -1,9 +1,19 @@
 define( [
     'qunit',
     'tests/mockup/mockup',
+    'osg/Light',
```
Comment (cedricpinson) [inline tests/osg/Matrix.js]: it's not the place to add more high level tests. Test with the same class name should unittest the class. For higher level test with sample scene or other should be in another directory.

**Comment by cedricpinson** [comment]
We can tomorrow to make the review together

**Comment by Kuranes** [comment]
@cedricpinson Pushed latest form, with its own benchmarks folder,

File: sources/osg/Matrix.js (hunk 2 above)
Comment (cedricpinson) [inline sources/osg/Matrix.js]: !fixed

File: sources/osg/Matrix.js (hunk 1 above)
Comment (cedricpinson) [inline sources/osg/Matrix.js]: !fixed

File: sources/osg/Node.js (hunk 3 above)
Comment (cedricpinson) [inline sources/osg/Node.js]: !fixed

File: tests/osg/Matrix.js (hunk 4 above)
Comment (cedricpinson) [inline tests/osg/Matrix.js]: !fixed

Comment (cedricpinson) [inline tests/osg/Matrix.js]: !fix

Comment (cedricpinson) [inline tests/osg/Matrix.js]: !fixed

File: sources/osg/Matrix.js [hunk 5]
```diff
@@ -1223,6 +1223,46 @@ define( [
             };
         } )(),
 
+        multScale: function ( x, y, z, m, res ) {
+
+            if ( !res ) {
```
Comment (cedricpinson) [inline sources/osg/Matrix.js]: Remove extra test it should never be called without res !fix

Comment (Kuranes) [inline sources/osg/Matrix.js]: k, so we rename it preMultScale for consistency and we'll make the multScale with a matrix result != m .

Comment (Kuranes) [inline sources/osg/Matrix.js]: !fixed

File: sources/osg/Matrix.js [hunk 6]
```diff
@@ -1224,6 +1224,28 @@ define( [
             };
         } )(),
 
+        preMultScale: function ( x, y, z, m ) {
```
Comment (cedricpinson) [inline sources/osg/Matrix.js]: Check how it's done for preMultTranslate, the order argument is not the same, matrix argument is always first.
The other point should be maybe to change premulttranslate to use it without array but scalar instead.
Or maybe we can do it in another commit. But for consistent both method should work the same !fix

Comment (Kuranes) [inline sources/osg/Matrix.js]: !fixed
